api/src/main/java/bio/terra/tanagra/service/search/utils/RandomNumberGenerator.java

```diff
@@ -0,0 +1,60 @@
+package bio.terra.tanagra.service.search.utils;
+
+import java.util.Random;
+
+/**
+ * Utility method for generating random numbers. Supports testing mode where the generator is
+ * re-seeded for each test, to produce an identical sequence that can be used in expected/actual
+ * comparisons.
+ */
+public final class RandomNumberGenerator {
+  private RandomNumberGenerator() {}
+
+  // default un-seeded random number generator used in production
+  private static final Random UNSEEDED_RANDOM = new Random();
+  // seeded random number generator used during testing
+  private static Random seededRandom = new Random();
+
+  // system property name indicating that we should use a seeded random (set by test harness before
+  // each test method)
+  private static final String USE_RANDOM_SEED_SYSTEM_PROPERTY = "GENERATE_SQL_RANDOM_SEED";
+
+  // whether to use the seeded random (true when testing), or unseeded (false when not testing)
+  private static final boolean USE_SEEDED_RANDOM = useSeededRandom();
+
+  // hard-coded (arbitrary) number to re-seed the generator with at the beginning of each test
+  private static final long RANDOM_SEED = 2022;
+
+  /** Return a random number generator. Used in SQL query generation. */
+  public static Random getRandom() {
+    return USE_SEEDED_RANDOM ? getSeededRandom() : UNSEEDED_RANDOM;
+  }
+
+  /**
+   * Check if the system property is set. This method is called to set a final static boolean, so
+   * that we only need to check the system property once in the non-testing case, not each time we
+   * need a random number.
+   *
+   * <p>See See bio.terra.tanagra.testing.BaseSpringUnitTest for where this is property set before
+   * each test method.
+   */
+  private static boolean useSeededRandom() {
+    String useRandomSeedSysProp = System.getProperty(USE_RANDOM_SEED_SYSTEM_PROPERTY);
+    return useRandomSeedSysProp != null && !useRandomSeedSysProp.isEmpty();
+  }
+
+  /** Return a seeded random generator. */
+  private static Random getSeededRandom() {
+    synchronized (RandomNumberGenerator.class) {
```

Comment:
I don't use much of the Java synchronzied keyword, so I may be missing something basic. What is the benefit of using a synchronized statement here vs making `getSeededRandom` a synchronized method?

Reply:
I originally defined the whole method as `synchronized`, but static analysis [prompted](https://pmd.github.io/pmd-6.31.0/pmd_rules_java_multithreading.html#avoidsynchronizedatmethodlevel) me to change it to a `synchronized` block instead so that we're not over-synchronizing in case of future additions to the method.